x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/AbstractPhysicalOperationProviders.java

```diff
@@ -138,7 +140,12 @@ else if (aggregatorMode == AggregatorMode.INITIAL || aggregatorMode == Aggregato
                 // knows that it should use a categorizing block hash.
                 // E.g. for `... BY groupAttribute = CATEGORIZE(sourceGroupAttribute)`, the `groupAttribute` will have type `Category`.
                 // We should put that info into the GroupSpec by other means.
-                groupSpecs.add(new GroupSpec(groupInput == null ? null : groupInput.channel(), groupAttribute));
+                // groupSpecs.add(new GroupSpec(groupInput == null ? null : groupInput.channel(), groupAttribute));
+
+                var groupSpec = Alias.unwrap(group) instanceof GroupingFunction groupingFunction
+                    ? new GroupSpec(groupInput == null ? null : groupInput.channel(), groupAttribute, groupingFunction)
+                    : new GroupSpec(groupInput == null ? null : groupInput.channel(), groupAttribute, null);
```

Comment:
Maybe we can also just expand the definition of `GroupSpec` to not rely on an attribute:

From
```
private record GroupSpec(Integer channel, Attribute attribute)
```
to
```
private record GroupSpec(Integer channel, NamedExpression groupExpression)
```

Or, if that's too annoying to wire in, we can also just add a boolean here to indicate whether or not we categorize on this group based on whether or not the grouping function is `instanceOf Categorize`.

Reply:
The attribute is being used in other place (`ordinalGroupingOperatorFactory`). So I left both for now: `Integer, Attribute, Expression (Instead of GroupingFunction)`.
No strong opinions here, so I would leave it that way for now.

About storing a boolean instead of the expression, I feel like the expression is as simple, and is not "too specific" for categorization